### PR TITLE
[WIP] ocaml: update to 5.1.1.

### PR DESCRIPTION
**Pull request by leahneukirchen**
Broken packages:
- [x] supermin
- [x] coccinelle
- [x] coq
	- [x] mathcomp
- [x] xen
- [x] zeroinstall

#### Testing the changes
- I tested the changes in this PR: **briefly**

### CODE REVIEW

**Comment by leahneukirchen** [comment]
Waiting for 5.1.

**Comment by kit-ty-kate** [comment]
> Waiting for 5.1.

5.1.0 is now released: https://discuss.ocaml.org/t/ocaml-5-1-0-released/13021

File: srcpkgs/ocaml/template [hunk 1]
```diff
@@ -1,26 +1,21 @@
...
-version=4.14.0
-revision=2
+version=5.0.0
+revision=1
 build_style="gnu-configure"
 configure_args="--with-pic --libdir=/usr/lib/ocaml"
 make_build_target="world.opt"
-makedepends="ncurses-devel libX11-devel"
-short_desc="Main implementation of the Caml language"
+makedepends="ncurses-devel"
```
Comment (kit-ty-kate) [inline srcpkgs/ocaml/template:8]:
```suggestion
makedepends="libzstd-devel"
```

**Comment by leahneukirchen** [comment]
Coq now builds, other stuff is still broken. :<

File: srcpkgs/ocaml/template [hunk 2]
```diff
@@ -1,26 +1,21 @@
...
-version=4.14.0
-revision=2
+version=5.1.0
+revision=1
 build_style="gnu-configure"
 configure_args="--with-pic --libdir=/usr/lib/ocaml"
 make_build_target="world.opt"
-makedepends="ncurses-devel libX11-devel"
-short_desc="Main implementation of the Caml language"
+makedepends="libzstd-devel"
```
Comment (kit-ty-kate) [inline srcpkgs/ocaml/template:8]: I don’t know void’s package system’s syntax but from the error messages it looks like the library is not kept after installation (this is currently linked dynamically so it should be installed at runtime too)

File: srcpkgs/ocaml/template [hunk 3]
```diff
@@ -1,26 +1,21 @@
...
+short_desc="Main implementation of the Objective Caml language"
```
Comment (kit-ty-kate) [inline srcpkgs/ocaml/template:9]: The language was simply renamed OCaml a long time ago
```suggestion
short_desc="Main implementation of the OCaml language"
```

File: srcpkgs/coq/template [hunk 4]
```diff
@@ -1,8 +1,9 @@
 # Template file for 'coq'
 pkgname=coq
 version=8.18.0
-revision=1
-hostmakedepends="camlp5 dune ocaml ocaml-findlib ocaml-num ocaml-zarith"
+revision=2
+hostmakedepends="dune ocaml ocaml-findlib ocaml-num ocaml-zarith"
+makedepends="libzstd-devel"
```
Comment (kit-ty-kate) [inline srcpkgs/coq/template:6]:
```suggestion
```

File: srcpkgs/ocaml/template (hunk 2 above)
Comment (leahneukirchen) [reply srcpkgs/ocaml/template:8]: The problem is that it's also needed to be linked against.

File: srcpkgs/coq/template (hunk 4 above)
Comment (kit-ty-kate) [reply srcpkgs/coq/template]: YIKES... NEVERMIND https://github.com/ocaml/ocaml/issues/12562#issuecomment-1721589602

**Review by kit-ty-kate — COMMENTED** [review]
5.1.1~rc1 is out so you should be able to test it (the final release should be available this week)

**Comment by MIvanchev** [comment]
Hey, I was working on this in parallel because I wasn't aware of this MR, why did the i686 job fail? I got it to run OK with `make world`. I was trying to get Haxe to build (would be a long process) and noticed other distros provide 2 packages: ocaml and ocaml-compiler-libs (see https://gitlab.archlinux.org/archlinux/packaging/packages/ocaml/-/blob/main/PKGBUILD?ref_type=heads for Arch). Maybe we should do the same?

**Comment by kit-ty-kate** [comment]
> why did the i686 job fail?

I don't remember the exact error but based on experience this is surely linked to the fact that OCaml >= 5.0 dropped support for native (bytecode still works) compilation on x86_32 and arm32 (ocamlc is still present, but ocamlopt is not)

> This is relevant because I got an error about missing compiler libs while trying stuff out.

Which architecture was this on? This is pretty weird because `make install` is supposed to install compiler-libs unconditionally.

**Comment by MIvanchev** [comment]
> I don't remember the exact error but based on experience this is surely linked to the fact that OCaml >= 5.0 dropped support for native (bytecode still works) compilation on x86_32 and arm32 (ocamlc is still present, but ocamlopt is not)

No, maybe it was something transient, the job should work now: https://github.com/void-linux/void-packages/actions/runs/7703464808/job/20993902940?pr=48430, maybe just re-run it.

> Which architecture was this on? This is pretty weird because `make install` is supposed to install compiler-libs unconditionally.

Hmm yeah, you are right, my job generated them:
https://github.com/void-linux/void-packages/actions/runs/7703464808/job/20993902799?pr=48430

I guess it wasn't the case previously or other distros just prefer separate packages for easier updates.

**Comment by kit-ty-kate** [comment]
> I guess it wasn't the case previously or other distros just prefer separate packages for easier updates.

This is the case since OCaml 4.00.0 apparently (released in 2012) so I'm guessing this is from a time before that.

**Comment by leahneukirchen** [comment]
The main blocker for this are packages which don't build on Ocaml 5.  You could help check the list and see if there are upstream fixes.

**Comment by MIvanchev** [comment]
> The main blocker for this are packages which don't build on Ocaml 5. You could help check the list and see if there are upstream fixes.

Good idea, let me just reopen my MR because of the convenience of CI :))

**Comment by MIvanchev** [comment]
Something bizarre is happening: when installed as dependency, `ocamlfind query <something>` reports `cannot find file 'query'` on i686. Prior to the installation it works if invoked through `${DESTDIR}/usr/bin/ocamlfind query <something>`. Investigating...

**Comment by MIvanchev** [comment]
Just want to report I managed to get everything to build so you might copy the changes from my branch and inspect the results. Some info:

* OCaml 5 doesn't have a native code compiler for 32 bit systems, the patches mainly deal with that
* Our COQ checks are not testing according to the devs, see https://github.com/coq/coq/issues/18613. If we do the real unit tests, we might not need the patch I provided to disable coqide builds
* Z3 installs a SO file in /usr/lib/stubfiles, might need to be moved

Maybe someone else can address these because I'm a  bit exhausted :smile:

**Comment by leahneukirchen** [comment]
Note to self: supermin 5.3.4 should work, xen should work, coccinelle can
be updated to a snapshot, 0install I'd remove unless someone finds a patch
(cc @mobinmob).

**Comment by kit-ty-kate** [comment]
> 0install I'd remove unless someone finds a patch

what is the failure with 0install? It works in opam (even with OCaml 5.2)

**Comment by MIvanchev** [comment]
@leahneukirchen You can fix the supermin musl build with

```
if [ "$XBPS_TARGET_LIBC" = musl ]; then
	makedepends+=" musl-fts-devel"
	LDFLAGS="-lfts"
fi
```
But I still have no clue why the tests are failing.

**Comment by MIvanchev** [comment]
OK so from what I see here the tests expect specific distributions and package managers so IHMO we have to skip some and patch the others.

**Comment by leahneukirchen** [comment]
@MIvanchev thanks, added.

**Comment by mobinmob** [comment]
@leahneukirchen Thanks for the mention and the fix.

**Pull request merged**